main.cpp

```diff
@@ -38,8 +40,289 @@
+        }
+        case 1: { // Mutate data
+            size_t original_data_size = data.size();
+            // Make sure we have a buffer that's large enough for mutation
+            data.resize(std::max(data.size(), data_maxlen));
+
+            size_t new_len = LLVMFuzzerMutate(data.data(),
+                                                original_data_size,
+                                                data_maxlen);
+
+            // It ensures that all values remain valid 5-bit values
+            for (auto& val : data) {
+                val &= 0x1F;
+            }
+            break;
+        }
+    }
+
+    if (!hrp.empty()) {
+        for (const char& c : hrp) {
+            if (c >= 'A' && c <= 'Z') {
+                return initial_input(fuzz_data, max_size);
+            }
+        }
+    }
+
+    std::string output = bech32::Encode(bech32::Encoding::BECH32, hrp, data);
+
+    if (output.length() > max_size) {
+        return initial_input(fuzz_data, max_size);
+    }
+
+    std::memcpy(fuzz_data, output.data(), output.length());
+    return output.length();
+}
+
+static std::vector<uint8_t> insert_part(std::span<const uint8_t> in1,
+                                      std::span<const uint8_t> in2,
+                                      size_t max_out_size)
+{
+    size_t in1_size = in1.size();
+    size_t in2_size = in2.size();
+    std::vector<uint8_t> out;
+
+    if (in1_size >= max_out_size)
+        return out;
+    if (in1_size == 0 || in2_size == 0)
+        return out;
+
+    size_t max_insert_size = max_out_size - in1_size;
+    if (max_insert_size > in2_size) max_insert_size = in2_size;
+    size_t insert_begin = std::rand() % in1_size;
+    size_t insert_size = (std::rand() % max_insert_size) + 1;
+
+    size_t in2_begin = std::rand() % (in2_size - insert_size + 1);
+
+    size_t total_size = in1_size + insert_size;
+    out.reserve(total_size);
+
+    out.insert(out.end(), in1.begin(), in1.begin() + insert_begin);
+    out.insert(out.end(), in2.begin() + in2_begin, in2.begin() + in2_begin + insert_size);
+    out.insert(out.end(), in1.begin() + insert_begin, in1.end());
+
+    return out;
+}
+
+static std::vector<uint8_t> overwrite_part(std::span<const uint8_t> in1,
+                                          std::span<const uint8_t> in2,
+                                          size_t max_out_size)
+{
+    std::vector<uint8_t> out;
+    out.reserve(max_out_size);
+
+    if (in1.empty() || in2.empty()) {
+        return out;
+    }
+
+    // Copy in1 to out first (limited by max_out_size)
+    size_t in1_bytes = std::min(in1.size(), max_out_size);
+    out.insert(out.end(), in1.begin(), in1.begin() + in1_bytes);
+
+    // Choose overwrite position
+    if (!out.empty() && !in2.empty()) {
+        size_t pos = std::rand() % out.size();
+        size_t max_possible = std::min(in2.size(), out.size() - pos);
+        size_t overwrite_length = max_possible > 0 ? (std::rand() % max_possible) + 1 : 0;
```

Comment:
AFAICT `max_possible` cannot be 0 here.

We already checked that `in2.size() > 0`, and `out.size() - pos` is always at least 1.

Reply:
Done